### ISSUE TITLE
Warning: a promise was created in a handler at 

### ISSUE THREAD

**Issue by jcald1** [OptimalBits/bull#518]
Using 3.0.0-alpha.2 on Node 6.9.1 on Ubuntu.  I get the following warnings.  I'm not getting these warnings with v2.2.6

(node:3215) Warning: a promise was created in a handler at /usr/apps/das/das_rtc/nodejs/TestConcurrency/node_modules/bull/lib/queue.js:811:5 but was not returned from it, see http://goo.gl/rRqMUw
    at new Promise (/usr/apps/das/das_rtc/nodejs/TestConcurrency/node_modules/bluebird/js/release/promise.js:79:10)

The warnings are printed even before any messages are placed on the queue for this code to process:

```
var request = require('request');
var sleep = require('sleep');
var Queue = require('bull');
var queue = Queue('testqueue', 6379, '127.0.0.1');

queue.process(function (job, done) {
    console.log('**Processing Queue Message: ' + JSON.stringify(job.data));
    console.log('Sending HTTP messaeg: ' + JSON.stringify(job.data));

    request('http://localhost:8080/respond1', function (error, response, body) {
        console.log('****In HTTP Callback');
        //console.log('error:', error); // Print the error if one occurred
        //console.log('statusCode:', response && response.statusCode); // Print the response status code if a response was received
        console.log('body:', body); // Print the HTML for the Google homepage.

        done();
    });
});
```

**Comment by bradvogel** [comment]
Add `return null;` as the last line of the function you pass to `process()`, per the link (http://goo.gl/rRqMUw)

**Comment by manast** [comment]
@jcald1 would you mind to test if this issue exists in 3.0.0-rc1?

**Comment by andrewburgess** [comment]
I'm seeing this warning when using delayed jobs with a crontab, coming from https://github.com/OptimalBits/bull/blob/f19632a2a909be327e6faa27a52c92649d6a16af/lib/queue.js#L661

**Comment by jcald1** [comment]
@manast, I'm not seeing the warnings any longer with  3.0.0-rc1

**Comment by fcarreiro** [comment]
I'm seeing this with delayed jobs with 3.0.0-rc2

**Comment by manast** [comment]
It is difficult to fix without being able to reproduce. Lets keep this open until somebody gets more information. At least we can say that the error is harmless, just a warning from bluebird where a error handling is missing somewhere.

**Comment by fcarreiro** [comment]
@manast I finally made some time to create an example to reproduce this bug.
I'm attaching a [bull-promises.zip](https://github.com/OptimalBits/bull/files/1150253/bull-promises.zip) file with

 * docker-compose.yml
 * Dockerfile
 * package.json
 * app.js

You can see all the precise versions there (`bull 3.0.0-rc.4` and node 8).

You are free to choose not to use Docker to run the example, but I'll write the instructions to run it with Docker:

(1) `docker-compose build`
(2) `docker-compose run --rm worker node app.js seed`
```
Creating network "bullpromises_default" with the default driver
Creating bull-promises-redis ... 
Creating bull-promises-redis ... done
Hit CTRL+C to stop
Seeding 1000 delayed messages...
Done, exiting.
```
(3) `docker-compose run --rm worker node app.js recover`
```
Starting bull-promises-redis ... done
Hit CTRL+C to stop
Recovering messages, processing...
(node:1) Warning: a promise was created in a handler at home/worker/node_modules/bull/lib/queue.js:279:17 but was not returned from it, see http://goo.gl/rRqMUw
    at new Promise (/home/worker/node_modules/bluebird/js/release/promise.js:79:10)
(node:1) Warning: a promise was created in a handler at home/worker/node_modules/bull/lib/queue.js:750:15 but was not returned from it, see http://goo.gl/rRqMUw
    at new Promise (/home/worker/node_modules/bluebird/js/release/promise.js:79:10)
(node:1) Warning: a promise was created in a handler at home/worker/node_modules/bull/lib/queue.js:750:15 but was not returned from it, see http://goo.gl/rRqMUw
    at new Promise (/home/worker/node_modules/bluebird/js/release/promise.js:79:10)
(node:1) Warning: a promise was created in a handler at home/worker/node_modules/bull/lib/queue.js:750:15 but was not returned from it, see http://goo.gl/rRqMUw
    at new Promise (/home/worker/node_modules/bluebird/js/release/promise.js:79:10)
(node:1) Warning: a promise was created in a handler at home/worker/node_modules/bull/lib/queue.js:750:15 but was not returned from it, see http://goo.gl/rRqMUw
    at new Promise (/home/worker/node_modules/bluebird/js/release/promise.js:79:10)
(node:1) Warning: a promise was created in a handler at home/worker/node_modules/bull/lib/queue.js:750:15 but was not returned from it, see http://goo.gl/rRqMUw
    at new Promise (/home/worker/node_modules/bluebird/js/release/promise.js:79:10)
(node:1) Warning: a promise was created in a handler at home/worker/node_modules/bull/lib/queue.js:750:15 but was not returned from it, see http://goo.gl/rRqMUw
    at new Promise (/home/worker/node_modules/bluebird/js/release/promise.js:79:10)
(node:1) Warning: a promise was created in a handler at home/worker/node_modules/bull/lib/queue.js:750:15 but was not returned from it, see http://goo.gl/rRqMUw
    at new Promise (/home/worker/node_modules/bluebird/js/release/promise.js:79:10)
(node:1) Warning: a promise was created in a handler at home/worker/node_modules/bull/lib/queue.js:750:15 but was not returned from it, see http://goo.gl/rRqMUw
    at new Promise (/home/worker/node_modules/bluebird/js/release/promise.js:79:10)
(node:1) Warning: a promise was created in a handler at home/worker/node_modules/bull/lib/queue.js:750:15 but was not returned from it, see http://goo.gl/rRqMUw
    at new Promise (/home/worker/node_modules/bluebird/js/release/promise.js:79:10)
(node:1) Warning: a promise was created in a handler at home/worker/node_modules/bull/lib/queue.js:750:15 but was not returned from it, see http://goo.gl/rRqMUw
    at new Promise (/home/worker/node_modules/bluebird/js/release/promise.js:79:10)
(node:1) Warning: a promise was created in a handler at home/worker/node_modules/bull/lib/queue.js:750:15 but was not returned from it, see http://goo.gl/rRqMUw
    at new Promise (/home/worker/node_modules/bluebird/js/release/promise.js:79:10)
(node:1) Warning: a promise was created in a handler at home/worker/node_modules/bull/lib/queue.js:750:15 but was not returned from it, see http://goo.gl/rRqMUw
    at new Promise (/home/worker/node_modules/bluebird/js/release/promise.js:79:10)
(node:1) Warning: a promise was created in a handler at home/worker/node_modules/bull/lib/queue.js:750:15 but was not returned from it, see http://goo.gl/rRqMUw
    at new Promise (/home/worker/node_modules/bluebird/js/release/promise.js:79:10)
(node:1) Warning: a promise was created in a handler at home/worker/node_modules/bull/lib/queue.js:750:15 but was not returned from it, see http://goo.gl/rRqMUw
    at new Promise (/home/worker/node_modules/bluebird/js/release/promise.js:79:10)
(node:1) Warning: a promise was created in a handler at home/worker/node_modules/bull/lib/queue.js:750:15 but was not returned from it, see http://goo.gl/rRqMUw
    at new Promise (/home/worker/node_modules/bluebird/js/release/promise.js:79:10)
(node:1) Warning: a promise was created in a handler at home/worker/node_modules/bull/lib/queue.js:750:15 but was not returned from it, see http://goo.gl/rRqMUw
    at new Promise (/home/worker/node_modules/bluebird/js/release/promise.js:79:10)
(node:1) Warning: a promise was created in a handler at home/worker/node_modules/bull/lib/queue.js:750:15 but was not returned from it, see http://goo.gl/rRqMUw
    at new Promise (/home/worker/node_modules/bluebird/js/release/promise.js:79:10)
(node:1) Warning: a promise was created in a handler at home/worker/node_modules/bull/lib/queue.js:750:15 but was not returned from it, see http://goo.gl/rRqMUw
    at new Promise (/home/worker/node_modules/bluebird/js/release/promise.js:79:10)
(node:1) Warning: a promise was created in a handler at home/worker/node_modules/bull/lib/queue.js:750:15 but was not returned from it, see http://goo.gl/rRqMUw
    at new Promise (/home/worker/node_modules/bluebird/js/release/promise.js:79:10)
(node:1) Warning: a promise was created in a handler at home/worker/node_modules/bull/lib/queue.js:750:15 but was not returned from it, see http://goo.gl/rRqMUw
    at new Promise (/home/worker/node_modules/bluebird/js/release/promise.js:79:10)
(node:1) Warning: a promise was created in a handler at home/worker/node_modules/bull/lib/queue.js:750:15 but was not returned from it, see http://goo.gl/rRqMUw
    at new Promise (/home/worker/node_modules/bluebird/js/release/promise.js:79:10)
(node:1) Warning: a promise was created in a handler at home/worker/node_modules/bull/lib/queue.js:750:15 but was not returned from it, see http://goo.gl/rRqMUw
    at new Promise (/home/worker/node_modules/bluebird/js/release/promise.js:79:10)
(node:1) Warning: a promise was created in a handler at home/worker/node_modules/bull/lib/queue.js:750:15 but was not returned from it, see http://goo.gl/rRqMUw
    at new Promise (/home/worker/node_modules/bluebird/js/release/promise.js:79:10)
(node:1) Warning: a promise was created in a handler at home/worker/node_modules/bull/lib/queue.js:750:15 but was not returned from it, see http://goo.gl/rRqMUw
    at new Promise (/home/worker/node_modules/bluebird/js/release/promise.js:79:10)
(node:1) Warning: a promise was created in a handler at home/worker/node_modules/bull/lib/queue.js:750:15 but was not returned from it, see http://goo.gl/rRqMUw
    at new Promise (/home/worker/node_modules/bluebird/js/release/promise.js:79:10)
(node:1) Warning: a promise was created in a handler at home/worker/node_modules/bull/lib/queue.js:750:15 but was not returned from it, see http://goo.gl/rRqMUw
    at new Promise (/home/worker/node_modules/bluebird/js/release/promise.js:79:10)
(node:1) Warning: a promise was created in a handler at home/worker/node_modules/bull/lib/queue.js:750:15 but was not returned from it, see http://goo.gl/rRqMUw
    at new Promise (/home/worker/node_modules/bluebird/js/release/promise.js:79:10)
(node:1) Warning: a promise was created in a handler at home/worker/node_modules/bull/lib/queue.js:750:15 but was not returned from it, see http://goo.gl/rRqMUw
    at new Promise (/home/worker/node_modules/bluebird/js/release/promise.js:79:10)
[2017-07-15T10:51:46.889Z] Processed #109
(node:1) Warning: a promise was created in a handler at home/worker/node_modules/bull/lib/queue.js:750:15 but was not returned from it, see http://goo.gl/rRqMUw
    at new Promise (/home/worker/node_modules/bluebird/js/release/promise.js:79:10)
(node:1) Warning: a promise was created in a handler at home/worker/node_modules/bull/lib/queue.js:750:15 but was not returned from it, see http://goo.gl/rRqMUw
    at new Promise (/home/worker/node_modules/bluebird/js/release/promise.js:79:10)
(node:1) Warning: a promise was created in a handler at home/worker/node_modules/bull/lib/queue.js:750:15 but was not returned from it, see http://goo.gl/rRqMUw
    at new Promise (/home/worker/node_modules/bluebird/js/release/promise.js:79:10)
(node:1) Warning: a promise was created in a handler at home/worker/node_modules/bull/lib/queue.js:750:15 but was not returned from it, see http://goo.gl/rRqMUw
    at new Promise (/home/worker/node_modules/bluebird/js/release/promise.js:79:10)
(node:1) Warning: a promise was created in a handler at home/worker/node_modules/bull/lib/queue.js:750:15 but was not returned from it, see http://goo.gl/rRqMUw
    at new Promise (/home/worker/node_modules/bluebird/js/release/promise.js:79:10)
(node:1) Warning: a promise was created in a handler at home/worker/node_modules/bull/lib/queue.js:750:15 but was not returned from it, see http://goo.gl/rRqMUw
    at new Promise (/home/worker/node_modules/bluebird/js/release/promise.js:79:10)
(node:1) Warning: a promise was created in a handler at home/worker/node_modules/bull/lib/queue.js:750:15 but was not returned from it, see http://goo.gl/rRqMUw
    at new Promise (/home/worker/node_modules/bluebird/js/release/promise.js:79:10)
(node:1) Warning: a promise was created in a handler at home/worker/node_modules/bull/lib/queue.js:750:15 but was not returned from it, see http://goo.gl/rRqMUw
    at new Promise (/home/worker/node_modules/bluebird/js/release/promise.js:79:10)
(node:1) Warning: a promise was created in a handler at home/worker/node_modules/bull/lib/queue.js:750:15 but was not returned from it, see http://goo.gl/rRqMUw
    at new Promise (/home/worker/node_modules/bluebird/js/release/promise.js:79:10)
(node:1) Warning: a promise was created in a handler at home/worker/node_modules/bull/lib/queue.js:750:15 but was not returned from it, see http://goo.gl/rRqMUw
    at new Promise (/home/worker/node_modules/bluebird/js/release/promise.js:79:10)
(node:1) Warning: a promise was created in a handler at home/worker/node_modules/bull/lib/queue.js:750:15 but was not returned from it, see http://goo.gl/rRqMUw
    at new Promise (/home/worker/node_modules/bluebird/js/release/promise.js:79:10)
(node:1) Warning: a promise was created in a handler at home/worker/node_modules/bull/lib/queue.js:750:15 but was not returned from it, see http://goo.gl/rRqMUw
    at new Promise (/home/worker/node_modules/bluebird/js/release/promise.js:79:10)
(node:1) Warning: a promise was created in a handler at home/worker/node_modules/bull/lib/queue.js:750:15 but was not returned from it, see http://goo.gl/rRqMUw
    at new Promise (/home/worker/node_modules/bluebird/js/release/promise.js:79:10)
(node:1) Warning: a promise was created in a handler at home/worker/node_modules/bull/lib/queue.js:750:15 but was not returned from it, see http://goo.gl/rRqMUw
    at new Promise (/home/worker/node_modules/bluebird/js/release/promise.js:79:10)
(node:1) Warning: a promise was created in a handler at home/worker/node_modules/bull/lib/queue.js:750:15 but was not returned from it, see http://goo.gl/rRqMUw
    at new Promise (/home/worker/node_modules/bluebird/js/release/promise.js:79:10)
(node:1) Warning: a promise was created in a handler at home/worker/node_modules/bull/lib/queue.js:750:15 but was not returned from it, see http://goo.gl/rRqMUw
    at new Promise (/home/worker/node_modules/bluebird/js/release/promise.js:79:10)
(node:1) Warning: a promise was created in a handler at home/worker/node_modules/bull/lib/queue.js:750:15 but was not returned from it, see http://goo.gl/rRqMUw
    at new Promise (/home/worker/node_modules/bluebird/js/release/promise.js:79:10)
(node:1) Warning: a promise was created in a handler at home/worker/node_modules/bull/lib/queue.js:750:15 but was not returned from it, see http://goo.gl/rRqMUw
    at new Promise (/home/worker/node_modules/bluebird/js/release/promise.js:79:10)
(node:1) Warning: a promise was created in a handler at home/worker/node_modules/bull/lib/queue.js:750:15 but was not returned from it, see http://goo.gl/rRqMUw
    at new Promise (/home/worker/node_modules/bluebird/js/release/promise.js:79:10)
(node:1) Warning: a promise was created in a handler at home/worker/node_modules/bull/lib/queue.js:750:15 but was not returned from it, see http://goo.gl/rRqMUw
    at new Promise (/home/worker/node_modules/bluebird/js/release/promise.js:79:10)
(node:1) Warning: a promise was created in a handler at home/worker/node_modules/bull/lib/queue.js:750:15 but was not returned from it, see http://goo.gl/rRqMUw
    at new Promise (/home/worker/node_modules/bluebird/js/release/promise.js:79:10)
(node:1) Warning: a promise was created in a handler at home/worker/node_modules/bull/lib/queue.js:750:15 but was not returned from it, see http://goo.gl/rRqMUw
    at new Promise (/home/worker/node_modules/bluebird/js/release/promise.js:79:10)
(node:1) Warning: a promise was created in a handler at home/worker/node_modules/bull/lib/queue.js:750:15 but was not returned from it, see http://goo.gl/rRqMUw
    at new Promise (/home/worker/node_modules/bluebird/js/release/promise.js:79:10)
(node:1) Warning: a promise was created in a handler at home/worker/node_modules/bull/lib/queue.js:750:15 but was not returned from it, see http://goo.gl/rRqMUw
    at new Promise (/home/worker/node_modules/bluebird/js/release/promise.js:79:10)
(node:1) Warning: a promise was created in a handler at home/worker/node_modules/bull/lib/queue.js:750:15 but was not returned from it, see http://goo.gl/rRqMUw
    at new Promise (/home/worker/node_modules/bluebird/js/release/promise.js:79:10)
(node:1) Warning: a promise was created in a handler at home/worker/node_modules/bull/lib/queue.js:750:15 but was not returned from it, see http://goo.gl/rRqMUw
    at new Promise (/home/worker/node_modules/bluebird/js/release/promise.js:79:10)
(node:1) Warning: a promise was created in a handler at home/worker/node_modules/bull/lib/queue.js:750:15 but was not returned from it, see http://goo.gl/rRqMUw
    at new Promise (/home/worker/node_modules/bluebird/js/release/promise.js:79:10)
[2017-07-15T10:51:46.984Z] Processed #42
[2017-07-15T10:51:46.996Z] Processed #535
[2017-07-15T10:51:47.007Z] Processed #963
[2017-07-15T10:51:47.018Z] Processed #14
[2017-07-15T10:51:47.031Z] Processed #240
[2017-07-15T10:51:47.037Z] Processed #93
[2017-07-15T10:51:47.046Z] Processed #627
[2017-07-15T10:51:47.051Z] Processed #363
[2017-07-15T10:51:47.060Z] Processed #609
[2017-07-15T10:51:47.065Z] Processed #187
[...]
```

Hope this helps!

**Comment by zidsi** [comment]
Still here at v.3.0.0-rc.9

In my case trace looks like this

```javascript
(node:10624) Warning: a promise was created in a handler but was not returned from it, see http://goo.gl/rRqMUw
    at new Promise (project\node_modules\bluebird\js\release\promise.js:79:10)
    at Command.initPromise (project\node_modules\bull\node_modules\ioredis\lib\command.js:63:18)
    at new Command (project\node_modules\bull\node_modules\ioredis\lib\command.js:51:8)
    at Script.execute (project\node_modules\bull\node_modules\ioredis\lib\script.js:22:17)
    at Redis.updateDelaySet (project\node_modules\bull\node_modules\ioredis\lib\commander.js:164:20)
    at Object.updateDelaySet (project\node_modules\bull\lib\scripts.js:248:25)
    at delayUpdate (project\node_modules\bull\lib\queue.js:702:15)
    at Queue.updateDelayTimer (project\node_modules\bull\lib\queue.js:718:25)
    at project\node_modules\bull\lib\queue.js:299:17
From previous event:
    at project\node_modules\bull\lib\queue.js:297:56
    at runCallback (timers.js:666:20)
    at tryOnImmediate (timers.js:639:5)
    at processImmediate [as _immediateCallback] (timers.js:611:5)
From previous event:
    at Queue._initProcess (project\node_modules\bull\lib\queue.js:293:8)
    at Queue.process (project\node_modules\bull\lib\queue.js:514:15)
   ....
```

and looks like this patch of queue.js removes the warning and should not impact functionality...

```diff
@@ -298,6 +298,7 @@
         if(timestamp){
           _this.updateDelayTimer(timestamp);
         }
+        return null;
       });
     }).then(function(){
       //
@@ -305,6 +306,7 @@
       // This is necessary when redis connection is unstable, which can cause the pub/sub to fail
       //
       _this.guardianTimer = setGuardianTimer(_this);
+      return null;
     });
 
     this.errorRetryTimer = {};
```

**Comment by manast** [comment]
Ok, i will take a look into this.

**Comment by zidsi** [comment]
I guess one more _runaway promise_ at delayUpdate...

My diff to v.3.0.0-rc.9 now looks like this. Hope it helps.

```diff
--- a/<html>queue.js </html>
+++ b/<html><b>Current File</b></html>
@@ -298,6 +298,7 @@
         if(timestamp){
           _this.updateDelayTimer(timestamp);
         }
+        return null;
       });
     }).then(function(){
       //
@@ -305,6 +306,7 @@
       // This is necessary when redis connection is unstable, which can cause the pub/sub to fail
       //
       _this.guardianTimer = setGuardianTimer(_this);
+      return null;
     });
 
     this.errorRetryTimer = {};
@@ -706,10 +708,12 @@
           nextTimestamp = Number.MAX_VALUE;
         }
         _this.updateDelayTimer(nextTimestamp);
+        return null;
       }).catch(function(err){
         _this.emit('error', err, 'Error updating the delay timer');
       });
       _this.delayedTimestamp = Number.MAX_VALUE;
+      
     };
 
     if(delay){
```

**Comment by manast** [comment]
fixed in afe960be7354b417ece8a4236638ee3e9241a0d8
